### ISSUE TITLE
Build fails when adding PactWindows nuget package to .net core project

### ISSUE THREAD

**Issue by ardz** [pact-foundation/pact-net#130]
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB3021	Unable to copy file "C:\Users\bob\.nuget\packages\pactnet.windows\2.2.0\tools\pact-win32\lib\vendor\ruby\2.2.0\gems\pact-mock_service-2.6.0\lib\pact\mock_service\request_handlers\base_administration_request_handler.rb" to "bin\Debug\netcoreapp2.0\pact-win32\lib\vendor\ruby\2.2.0\gems\pact-mock_service-2.6.0\lib\pact\mock_service\request_handlers\base_administration_request_handler.rb". The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.	TestPact C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets	4406

**Comment by neilcampbell** [comment]
Hi @ardz 
Unfortunately that is a windows limitation, so there is not much I can do.
Doing a character count on both those paths listed is less than the character limit, so not sure what is going on there. Is the path to your project quite long? Can you try shorting the path where your source code is stored?

**Comment by ardz** [comment]
The interesting thing is, when you use a .net standard project and add the the nuget package the build doesn't fail with the above error.

However, the examples for pact net use xUnit test framework, and to use the xUnit test framework with VS2017 you need to make the project a .net core project.

:|

**Comment by neilcampbell** [comment]
Hmm, thats odd. You can use whatever test framework you like, PactNet is completely agnostic.

**Comment by awiesendangermdsol** [comment]
Just as follow up, this is an issue w/copying the ruby files. If you want to use PACT-NET, you have to make sure the path for your projects is limited. Not sure what a safe amount is. But I copied the workshop [example project](https://github.com/tdshipley/pact-workshop-dotnet-core-v1), and built it with out issues. Then I renamed a few folders in the path to long names, and got the same error building, with no other changes.

**Comment by neilcampbell** [comment]
Unfortunately it's a windows limitation, such a pain!

**Comment by mattfrear** [comment]
I understand that this is a windows limitation. But WTH is going on? Why is pact.net trying to copy a file to bin\Debug\netcoreapp2.1\pact-win32\lib\ruby\lib\ruby\gems\2.2.0\gems\bundler-1.9.9\lib\bundler\vendor\molinillo\lib\molinillo\modules\specification_provider.rb

Does it _really have to_ copy a file to lib\ruby\lib\ruby and then molinillo\lib\molinillo etc etc? Can't it be reorganised so that the folder structure isn't so damn deep?

**Comment by neilcampbell** [comment]
@mattfrear I hear you, I know it sucks! It's how the ruby dependency setup works in the Pact standalone core and unfortunately it cannot be changed. The only way forward is to move to using the rust core https://github.com/pact-foundation/pact-reference. I haven't had a chance to look into this, as it's quite a bit of work and will likely have other tradeoffs.

**Comment by mattfrear** [comment]
I don't know anything about Ruby, but could it be packed into an exe?

Found this https://github.com/pmq20/ruby-packer

Matt

**Comment by bethesque** [comment]
We've been using Travelling Ruby to create the package, but it's very out of date (only supports old version of Ruby). I'll have a look at Ruby Packer, thanks for the tip.

**Comment by bethesque** [comment]
I've had a quick play with it, and the linux packaging works nicely BUT the package is 15MB compressed as distinct from 9MB compressed. It's 21MB uncompressed. I assume you include the uncompressed version in the .net package @neilcampbell

**Comment by bethesque** [comment]
I'll give it a go. One thing that's going to complicate things is that currently, the standalone package allows us to group multiple entrance points (eg `pact-mock-service`, `pact-provider-verifier`, `pact-broker`). I'll have to create one single entrance point (`pact`) to delegate to each of the separate CLIs (which @mefellows was suggesting at one point anyway) I'll let you know how I go.

**Comment by neilcampbell** [comment]
Nice one, sounds great.

**Comment by bethesque** [comment]
Having trouble with windows. 

C:/__enclose_io_memfs__/local/lib/compiler/utils.rb:31:in `spawn': No such file or directory - nmake /f win32\Makefile.msc (Errno::ENOENT)

Does that mean it can't find nmake or Makefile.msc?

**Comment by bethesque** [comment]
If anyone wants to play along at home

https://github.com/pact-foundation/pact-ruby-cli/blob/master/appveyor.yml
https://ci.appveyor.com/project/pact-foundation/pact-ruby-cli

**Comment by bethesque** [comment]
Progress update - I'm hitting some ruby/windows issues with ruby-packer and I'm trying to get help from the author. For some reason the `rubyc` command that creates the executable is trying to load a "socket.so" file, but .so files aren't a windows Thing, so I'm not sure why that's happening.

**Comment by Matt-VL** [comment]
Do you have an update on this? Running into this issue on my machine currently.

**Comment by mattfrear** [comment]
Ruby compiler (aka ruby packer) which I suggested hasn't had any releases since 2017 so that might be a dead end.

https://github.com/pmq20/ruby-packer/releases

**Comment by Matt-VL** [comment]
@bethesque thanks for the update. Is there a way I can get updates on that progress? At this point I am dead in the water with the only workaround being to run my code on linux which would be a hassle.

**Comment by greghicks01** [comment]
@bethesque In relation to the last comments above, was there any progress with pact-js rust implementation?

**Comment by mefellows** [comment]
Hi @greghicks01, let's discuss in the #pact-net-development channel. I'll give a brief update there.

TL;DR - the Rust FFI bindings for the provider interface are ✅ , the FFI bindings for the consumer side are about 80% there (no request matching rules or generators on the headers, query string, path). But by the time the work is done to set up here, the Rust code may be ready.

See https://github.com/pact-foundation/pact-go/blob/feat/v3-ffi-verifier/v3/internal/native/mockserver/mock_server.go for the Golang raw FFI boundary.
